returnn/datasets/postprocessing.py

```diff
@@ -296,7 +298,73 @@ def _make_tensor_template_from_input(self, data_key: str) -> Tensor:
         return Tensor(data_key, dims=dims, dtype=dtype, sparse_dim=sparse_dim)
 
 
-class LaplaceOrdering(Callable[[Iterator[TensorDict]], Iterator[TensorDict]]):
+class SeqOrdering:
+    """Base class for defining seq ordering iterators compatible w/ the :class:`PostprocessingDataset`."""
+
+    def __init__(self, length_key: str):
+        """:param length_key: data key to determine the segment length from for ordering."""
+        self._length_key = length_key
+
+    def _get_seq_len(self, tdict: TensorDict) -> int:
+        """
+        :return: segment length of the segment in `tdict` as measured by `self._length_key` for comparison.
+        """
+        return tdict.data[self._length_key].raw_tensor.shape[0]
+
+
+class BucketOrdering(SeqOrdering, Callable[[Iterator[TensorDict]], Iterator[TensorDict]]):
+    """
+    Iterator compatible with :class:`PostprocessingDataset`'s ``map_seq_stream`` applying
+    an ordering that creates batches consisting of segments falling into a few distinct
+    length buckets.
+
+    To be composed with any custom data postprocessing logic via :class:`Sequential`.
+
+    See also :class:`LaplaceOrdering`.
+    """
+
+    def __init__(self, seq_lens: Sequence[int], length_key: str = "data"):
```

Comment:
For every max_seq_len, you should define the max_num_seqs. The option here should be more like:
```python
bucket_sizes: List[Tuple[int, int]]
```
representing the (max_seq_len,max_num_seqs) pair.